server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSetMethod.scala

```diff
@@ -0,0 +1,133 @@
+import org.apache.james.jmap.mail.{DestroyIds, EmailSet, EmailSetRequest, EmailSetResponse}
+import org.apache.james.jmap.model.CapabilityIdentifier.CapabilityIdentifier
+import org.apache.james.jmap.model.DefaultCapabilities.{CORE_CAPABILITY, MAIL_CAPABILITY}
+import org.apache.james.jmap.model.Invocation.{Arguments, MethodName}
+import org.apache.james.jmap.model.SetError.SetErrorDescription
+import org.apache.james.jmap.model.{Capabilities, Invocation, SetError, State}
+import org.apache.james.mailbox.model.{DeleteResult, MessageId}
+import org.apache.james.mailbox.{MailboxSession, MessageIdManager}
+import org.apache.james.metrics.api.MetricFactory
+import play.api.libs.json.{JsError, JsSuccess}
+import reactor.core.scala.publisher.{SFlux, SMono}
+import reactor.core.scheduler.Schedulers
+
+import scala.jdk.CollectionConverters._
+
+case class MessageNotFoundExeception(messageId: MessageId) extends Exception
+
+case class DestroyResults(results: Seq[DestroyResult]) {
+  def destroyed: Option[DestroyIds] = {
+    Option(results.flatMap({
+      result => result match {
+        case result: DestroySuccess => Some(result.messageId)
+        case _ => None
+      }
+    }).map(EmailSet.asUnparsed))
+      .filter(_.nonEmpty)
+      .map(DestroyIds)
+  }
+
+  def notDestroyed: Option[Map[UnparsedMessageId, SetError]] = {
+    Option(results.flatMap({
+      result => result match {
+        case failure: DestroyFailure => Some(failure)
+        case _ => None
+      }
+    })
+      .map(failure => (failure.unparsedMessageId, failure.asMessageSetError))
+      .toMap)
+      .filter(_.nonEmpty)
+  }
+}
+
+object DestroyResult {
+  def from(deleteResult: DeleteResult): DestroyResult = {
+    val notFound = deleteResult.getNotFound.asScala
+
+    deleteResult.getDestroyed.asScala
+      .headOption
+      .map(DestroySuccess)
+      .getOrElse(DestroyFailure(EmailSet.asUnparsed(notFound.head), MessageNotFoundExeception(notFound.head)))
+  }
+}
+
+trait DestroyResult
+case class DestroySuccess(messageId: MessageId) extends DestroyResult
+case class DestroyFailure(unparsedMessageId: UnparsedMessageId, e: Throwable) extends DestroyResult {
+  def asMessageSetError: SetError = e match {
+    case e: IllegalArgumentException => SetError.invalidArguments(SetErrorDescription(s"$unparsedMessageId is not a messageId: ${e.getMessage}"))
+    case e: MessageNotFoundExeception => SetError.notFound(SetErrorDescription(s"Cannot find message with messageId: ${e.messageId.serialize()}"))
+    case _ => SetError.serverFail(SetErrorDescription(e.getMessage))
+  }
+}
+
+class EmailSetMethod @Inject()(serializer: EmailSetSerializer,
+                               messageIdManager: MessageIdManager,
+                               messageIdFactory: MessageId.Factory,
+                               val metricFactory: MetricFactory,
+                               val sessionSupplier: SessionSupplier) extends MethodRequiringAccountId[EmailSetRequest] {
```

Comment:
Why some have val and some others don't ?

Reply:
val will implicitly generate getter for those fields, which are required because EmailSetMethod is extending MethodRequiringAccountId